src/abstractdataframe/iteration.jl

```diff
@@ -159,15 +161,28 @@ julia> sum.(eachcol(df))
 """
 eachcol(df::AbstractDataFrame) = DataFrameColumns(df)
 
+Base.IteratorSize(::Type{<:DataFrameColumns}) = Base.HasShape{1}()
 Base.size(itr::DataFrameColumns) = (size(parent(itr), 2),)
-Base.IndexStyle(::Type{<:DataFrameColumns}) = Base.IndexLinear()
 
-@inline function Base.getindex(itr::DataFrameColumns, j::Int)
-    @boundscheck checkbounds(itr, j)
-    @inbounds parent(itr)[!, j]
+function Base.size(itr::DataFrameColumns, d::Integer)
+    d < 1 && throw(ArgumentError("dimension out of range"))
+    return d == 1 ? size(itr)[1] : 1
 end
 
-Base.getindex(itr::DataFrameColumns, j::Symbol) = parent(itr)[!, j]
+Base.length(itr::DataFrameColumns) = size(itr)[1]
+Base.eltype(::Type{<:DataFrameColumns}) = AbstractVector
+Base.firstindex(itr::DataFrameColumns) = 1
+Base.lastindex(itr::DataFrameColumns) = length(itr)
+Base.iterate(itr::DataFrameColumns, i::Integer=1) =
+    i <= length(itr) ? (itr[i], i + 1) : nothing
+Base.@propagate_inbounds Base.getindex(itr::DataFrameColumns, idx::ColumnIndex) =
+    parent(itr)[!, idx]
+Base.@propagate_inbounds Base.getindex(itr::DataFrameColumns, idx::MultiColumnIndex) =
+    eachcol(parent(itr)[!, idx])
```

Comment:
Your questions prompted me to discuss this line. When using `MultiColumnIndex` we create a new data frame, but **do not** copy its columns (so it contains aliases). It is faster this way, and consistent with `getindex` for `ColumnIndex` but it is probably a bit unsafe. Do you think we should mention this behaviour somewhere?

Reply:
Actually I have decided to change this to produce a `view` as it is cleaner.